google/resumable_media/_helpers.py

```diff
@@ -162,23 +162,43 @@ def wait_and_retry(func, get_status_code, retry_strategy):
     Returns:
         object: The return value of ``func``.
     """
-    response = func()
-    if get_status_code(response) not in RETRYABLE:
-        return response
-
     total_sleep = 0.0
     num_retries = 0
     base_wait = 0.5  # When doubled will give 1.0
-    while retry_strategy.retry_allowed(total_sleep, num_retries):
+
+    # Set the retriable_exception_type if possible. We expect requests to be
+    # present here and the transport to be using requests.exceptions errors,
+    # but due to loose coupling with the transport layer we can't guarantee it.
+    try:
+        import requests
+
+        retriable_exception_type = requests.exceptions.ConnectionError
+    except ImportError:
+        retriable_exception_type = None
+
+    while True:  # return on success or when retries exhausted.
+        error = None
+        try:
+            response = func()
+        except retriable_exception_type as e:
+            error = e
+        else:
+            if get_status_code(response) not in RETRYABLE:
+                return response
+
+        if not retry_strategy.retry_allowed(total_sleep, num_retries):
+            # Retries are exhausted and no acceptable response was received. Raise the
+            # retriable_error or return the unacceptable response.
+            if error:
+                raise error
+
+            return response
+
         base_wait, wait_time = calculate_retry_wait(base_wait, retry_strategy.max_sleep)
+
         num_retries += 1
         total_sleep += wait_time
         time.sleep(wait_time)
-        response = func()
-        if get_status_code(response) not in RETRYABLE:
```

Comment:
It's interesting that this now happens before the wait rather than after. Seems like a correct change though?

Reply:
The reason here is simply that the original code called func() both before the while loop (zeroth retry) and inside it (first retry and beyond). The new one only has one call to func(), inside the loop. Any retry strategy stuff only fires up after the first call is complete in both designs, but this one unifies the func() calls.